### ISSUE TITLE
Text overlapping in Vim over SSH with colors enabled

### ISSUE THREAD

**Issue by sryze** [git-for-windows/git#1520]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.14.1.windows.1
built from commit: 82d9b3f3b2407b52251620597d4b14933685459d
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.16299.248]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

None

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash (not mintty)

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

I'm having this problem with a server running CentOS 7. I wasn't able to reproduce this with a local VirtualBox VM. 

Basically I SSH into it and open a file in vim, and after opening the text inside the file gets mixed withe the text from previous console output or even with the text from my Windows bash session.

```
ssh some-host
vim some-file
```

I think having colors in vim messes everything up. If I execute this after connecting via ssh and before opening vim:

```
export XTERM=xterm-old
```

the bug disappears. Also when I connect from terminal (gnome-terminal) on a Linux machine I don't see this issue.

Another workaround that I used is select all text in the file in visual mode and then exit visual mode.

 - What did you expect to occur after running these commands?

Content of the file

 - What actually happened instead?

File content is intermixed with console output

**Comment by sryze** [comment]
I don't know if this is the right place to report Git Bash related bugs, but I haven't found any other suitable place. I actually found another bug that makes `less` crash, show I also create it under this repository?

**Comment by dscho** [comment]
@sryze this may be caused by Git for Windows shipping only partial packages (e.g. only a subset of terminfo, to save on bandwidth).

Could you verify that this also happens in a regular [MSYS2](https://msys2.github.io/) setup?

**Comment by sryze** [comment]
MSYS2 uses mintty, ssh + vim works fine there. If I Git Bash with mintty it also works. Is there any way to run MSYS2 without mintty? I tried running usr/bin/sh.exe from inside cmd shell but sh wasn't able to find any commands (like ls).

I would use mintty with Git Bash but unfortunately ConEmu doesn't support that so I would have run it separately.

**Comment by sryze** [comment]
Hmm, after removing `TERM="xterm-256color"` from my .bashrc I haven't experienced this bug. I'll do some more testing and close this if it doesn't repeat.

**Comment by dscho** [comment]
@sryze there has been a recent bugfix in ncurses which avoided problems when `TERM` was set to a value for which no terminfo can be found (and Git for Windows saves on bandwidth and excludes most of the terminfo offered by MSYS2).

Could you try the latest snapshot installer at https://wingit.blob.core.windows.net/files/index.html?

**Comment by sryze** [comment]
I installed the version from 13 Mar 2018 and unfortunately the bug is present in it. The workaround that I mentioned above is still working though (when I don't set `TERM` and it's set to `cygwin` automatically).

**Comment by dscho** [comment]
@sryze just to be clear: *which* bug are you talking about? The `less.exe` crash? Or the overlapping text?

**Comment by sryze** [comment]
The one related to overlapping text, less seems to work more or less now.  With less the bug was that if you do `export TERM=` without a value it would crash, now it doesn't crash and shows an error instead - `WARNING: terminal is not fully functional`.

**Comment by dscho** [comment]
> With less the bug was that if you do export TERM= without a value it would crash, now it doesn't crash and shows an error instead - WARNING: terminal is not fully functional.

Yes, that's the bug fix I referenced.

The overlapping text may be caused by Git for Windows shipping with terminfo for `cygwin` that might be incorrect? I do not know, I am not an expert in terminfo issues. Maybe the `cygwinB19` and `cygwinDBG` files are required, too? (You could copy them from https://github.com/git-for-windows/git-sdk-64/tree/master/usr/lib/terminfo/63 to the `/usr/share/terminfo/63/` and `/usr/lib/terminfo/63/` subdirectories.)

Or the terminfo is simply incorrect. I am *far* from understanding what they do, but it may be an issue you can figure out from reading the output of `infocmp cygwin` in a Git for Windows SDK?

**Comment by sryze** [comment]
Well, I'm surely even father from understanding that because I have no idea what terminfo is. I'll try what you suggested, maybe something will come out of it.

**Comment by dscho** [comment]
@sryze I am far from an expert, but... terminfo is a way to describe what byte sequences to use to do certain things such as: move the cursor to a given column and row in the current terminal window. These things are different between different terminals, and the `TERM` variable is supposed to tell you which terminfo to use. That's as far as my knowledge goes, though ;-)

**Comment by dscho** [comment]
> I'll try what you suggested, maybe something will come out of it.

Did you try? Did something come out of it? Please do not let the time I spent trying to help you go to waste.

**Comment by sryze** [comment]
I did not, sorry, I can't reproduce this bug anymore, so I'm closing this.
